linera-core/src/client/mod.rs

```diff
@@ -422,6 +423,53 @@ where
     }
 }
 
+// Tracks which chain_id<>block_height pairs have been successfully processed.
+struct MultichainTracker {
+    // Starting tracker index.
+    tracker: u64,
+    // Log of received certificates by the remote node.
+    remote_node_received_log: Vec<ChainAndHeight>,
+    // Track which certificates have been successfully processed.
+    // We will use that information later to update the tracker.
+    successfuly_processed: BTreeMap<ChainId, Vec<BlockHeight>>,
```

Comment:
Would be more efficient to make this a set instead of a vector.

Reply:
Yes, I will be iterating on the data structures before finalising the PR. Ideally we could store itself `ChainAndHeight` in a `HashSet` but it doesn't implement `Hash`. Neither it implements `Ord` so we can't store in a `BTreeSet`.

Reply:
Right; I guess we could derive either for it.

Reply:
OK, I think we can have a set after all – we're processing through the `remote_node_received_log` and just calling `contains` which doesn't remote the elements. So the same `ChaindAndHeight` elements will all return `true`.